### PR TITLE
Add aria roles and labels

### PR DESCRIPTION
**Pull request by marthacryan**
## References
Fixes the first part of #9491 to add basic roles and aria labels to various panels in JupyterLab. 

## Code changes
See issue for specific areas that are changed. The code only adds attributes to nodes of widgets. 
Todo:
- [x] Add translation

## User-facing changes
For non screenreaders, there are no user-facing changes. This gives more information that can be used for accessibility purposes. 

## Backwards-incompatible changes
None

### CODE REVIEW

**Comment by marthacryan** [comment]
@manfromjupyter I'll need to add translation to this PR for the aria labels, but otherwise if you have a chance, I think it's ready for you to take a look.

File: packages/apputils-extension/src/palette.ts [hunk 1]
```diff
@@ -103,6 +103,8 @@ export namespace Palette {
     const modalPalette = new ModalCommandPalette({ commandPalette: palette });
     let modal = false;
 
+    palette.node.setAttribute('role', 'region');
+    palette.node.setAttribute('aria-label', 'Command Palette Section');
```
Comment (goanpeca) [inline packages/apputils-extension/src/palette.ts:107]:
```suggestion
    palette.node.setAttribute('aria-label', trans.__('Command Palette Section'));
```

File: packages/apputils/src/mainareawidget.ts [hunk 2]
```diff
@@ -38,7 +38,11 @@ export class MainAreaWidget<T extends Widget = Widget>
     this.id = DOMUtils.createDomID();
```
Comment (goanpeca) [inline packages/apputils/src/mainareawidget.ts:38]: We could add a new optional parameter to the constructor for the translator.

File: packages/application/src/shell.ts [hunk 3]
```diff
@@ -192,8 +192,12 @@ export class LabShell extends Widget implements JupyterFrontEnd.IShell {
```
Comment (goanpeca) [inline packages/application/src/shell.ts:192]: We could add a new optional parameter to the constructor for the translator.

File: packages/debugger-extension/src/index.ts [hunk 4]
```diff
@@ -512,6 +512,9 @@ const main: JupyterFrontEndPlugin<void> = {
       restorer.add(sidebar, 'debugger-sidebar');
     }
 
+    sidebar.node.setAttribute('role', 'region');
+    sidebar.node.setAttribute('aria-label', 'Debugger section');
```
Comment (goanpeca) [inline packages/debugger-extension/src/index.ts:516]:
```suggestion
    sidebar.node.setAttribute('aria-label', trans.__('Debugger section'));
```

File: packages/extensionmanager-extension/src/index.ts [hunk 5]
```diff
@@ -64,6 +64,8 @@ const plugin: JupyterFrontEndPlugin<void> = {
 
     if (enabled && labShell) {
       view = createView();
+      view.node.setAttribute('role', 'region');
+      view.node.setAttribute('aria-label', 'Extension Manager section');
```
Comment (goanpeca) [inline packages/extensionmanager-extension/src/index.ts:68]:
```suggestion
      view.node.setAttribute('aria-label', trans.__('Extension Manager section'));
```

File: packages/extensionmanager-extension/src/index.ts [hunk 6]
```diff
@@ -80,6 +82,8 @@ const plugin: JupyterFrontEndPlugin<void> = {
               return;
             }
             view = view || createView();
+            view.node.setAttribute('role', 'region');
+            view.node.setAttribute('aria-label', 'Extension Manager section');
```
Comment (goanpeca) [inline packages/extensionmanager-extension/src/index.ts:86]:
```suggestion
            view.node.setAttribute('aria-label', trans.__('Extension Manager section'));
```

File: packages/filebrowser-extension/src/index.ts [hunk 7]
```diff
@@ -325,6 +325,9 @@ function activateBrowser(
 ): void {
   const trans = translator.load('jupyterlab');
   const browser = factory.defaultBrowser;
+  browser.node.setAttribute('role', 'region');
+  browser.node.setAttribute('aria-label', 'File Browser Section');
```
Comment (goanpeca) [inline packages/filebrowser-extension/src/index.ts:329]:
```suggestion
  browser.node.setAttribute('aria-label', trans.__('File Browser Section'));
```

File: packages/filebrowser/src/browser.ts [hunk 8]
```diff
@@ -86,6 +86,9 @@ export class FileBrowser extends Widget {
     this._trans = this.translator.load('jupyterlab');
     this._crumbs = new BreadCrumbs({ model, translator });
     this.toolbar = new Toolbar<Widget>();
+    // a11y
+    this.toolbar.node.setAttribute('role', 'navigation');
+    this.toolbar.node.setAttribute('aria-label', 'file browser');
```
Comment (goanpeca) [inline packages/filebrowser/src/browser.ts:91]:
```suggestion
    this.toolbar.node.setAttribute('aria-label', this._trans('file browser'));
```

File: packages/running-extension/src/index.ts [hunk 9]
```diff
@@ -58,6 +58,8 @@ function activate(
   running.id = 'jp-running-sessions';
   running.title.caption = trans.__('Running Terminals and Kernels');
   running.title.icon = runningIcon;
+  running.node.setAttribute('role', 'region');
+  running.node.setAttribute('aria-label', 'Running Sessions section');
```
Comment (goanpeca) [inline packages/running-extension/src/index.ts:62]:
```suggestion
  running.node.setAttribute('aria-label', trans._('Running Sessions section'));
```

File: packages/toc-extension/src/index.ts [hunk 10]
```diff
@@ -66,6 +66,9 @@ async function activateTOC(
   toc.title.icon = tocIcon;
   toc.title.caption = trans.__('Table of Contents');
   toc.id = 'table-of-contents';
+  toc.node.setAttribute('role', 'region');
+  toc.node.setAttribute('aria-label', 'Table of Contents Section');
```
Comment (goanpeca) [inline packages/toc-extension/src/index.ts:70]:
```suggestion
  toc.node.setAttribute('aria-label', trans.__('Table of Contents Section'));
```

**Comment by goanpeca** [comment]
Thanks for working on this @marthacryan :), looking good!

I left some inline comments. The `trans` or `this._trans` is available for most of the files changed. We need to add a couple of params on constructors for the remaining ones.

The only doubt I have is if the translator should be added as a separate parameter if the class already takes some sort of `IClassOptions` or as part of those options.

@blink1073 ?

**Comment by manfromjupyter** [comment]
> @manfromjupyter I'll need to add translation to this PR for the aria labels, but otherwise if you have a chance, I think it's ready for you to take a look.

@marthacryan Wow that was fast! Will test it within a couple hours here and include screenshots. Can test it after you make Gonzalo's changes too, but I doubt it will change anything for the screenreader.

**Comment by manfromjupyter** [comment]
@marthacryan, It was working but all of the sudden the Binder is not working due to I think the translation piece.. Might just be a simple syntax issue. See screenshot:
translationbundle

File: packages/application/src/shell.ts [hunk 11]
```diff
@@ -185,15 +196,25 @@ export class LabShell extends Widget implements JupyterFrontEnd.IShell {
...
 
+    const trans = ((options && options.translator) || nullTranslator).load(
+      'jupyterlab'
+    );
     const headerPanel = (this._headerPanel = new BoxPanel());
     const menuHandler = (this._menuHandler = new Private.PanelHandler());
+    menuHandler.panel.node.setAttribute('role', 'navigation');
+    menuHandler.panel.node.setAttribute(
+      'aria-label',
+      trans.__('main navigation')
```
Comment (manfromjupyter) [inline packages/application/src/shell.ts:212]:
```suggestion
      trans.__('main')
```
My bad in the ticket I sent you. The word navigation is repeated twice which is just annoying. If screenreaders go to that element by jumping to it, viewing next element, or viewing all navigations, "main" is all that is needed.
navigation of header nav labeled twice

**Comment by marthacryan** [comment]
@goanpeca I added an optional argument to `LabShell` for the translation but because of the way it's instantiated (as an argument to the `JupyterLab` object) I'm not sure how to access the `ITranslator` object (I'm accustomed to taking `ITranslator` as an argument to the `activate` function of an extension). Any advice?

File: packages/application/src/shell.ts [hunk 12]
```diff
@@ -213,12 +234,34 @@ export class LabShell extends Widget implements JupyterFrontEnd.IShell {
...
     leftHandler.sideBar.addClass('jp-mod-left');
+    leftHandler.sideBar.node.setAttribute('role', 'complementary');
+    leftHandler.sideBar.node.setAttribute(
+      'aria-label',
+      trans.__('main sidebar')
+    );
+    leftHandler.sideBar.contentNode.setAttribute('role', 'navigation');
+    leftHandler.sideBar.contentNode.setAttribute(
+      'aria-label',
+      trans.__('main sidebar navigation')
```
Comment (manfromjupyter) [inline packages/application/src/shell.ts:245]: Same thing for this one. My apologies - could have labeled my pictures better.
```suggestion
      trans.__('main sidebar')
```
main sidebar double nav

**Comment by goanpeca** [comment]
Hi @marthacryan, you mean here https://github.com/jupyterlab/jupyterlab/blob/master/packages/application/src/lab.ts#L29 ?

**Comment by marthacryan** [comment]
> Hi @marthacryan, you mean here https://github.com/jupyterlab/jupyterlab/blob/master/packages/application/src/lab.ts#L29 ?

Yes

**Comment by goanpeca** [comment]
Uhhh tricky :-\, @jasongrout @blink1073 how can we pass the Itranslator to the lab shell here?

https://github.com/jupyterlab/jupyterlab/blob/master/packages/application/src/lab.ts#L29 ?

**Comment by manfromjupyter** [comment]
@marthacryan 2 things I think are beyond the scope of this ticket but would like to note for you, the accessibility task force, and those that care I guess haha.
1. The two sidebar regions can't be accessed to date with a screereader because there is technically no element that is targetable within them. To prove this, by adding a single text character the regions suddenly kick in and I can access them with a screenreader. They aren't in a parent element that is targetable and no children text or element is in there.
example, I had to inject `"hi"` to gain focus into the container as a whole:
main sidebar double nav

2. Mary added the appropriate regions to the different "Manager" panels. Files, Running Sessions, Extensions, etc but each of these need cleanup as they use invalid semantic HTML (ex. `<header>` tags in place of `<H2>`'s, etc. Fixing these may affect the themes extensions and overall design so will need to tackle those section by section. But anyway, these act as their own regions so may because they have no text under them before said site header (banner) appears according to the html.

Screenshots:
Table of contents section reading the title as a site header
running sessions section
extensions region invalid html  region breaks

So anyway for both of these ordered list items it may not look like Mary's changes are working, but they're in there and are. The foundation is laid, we just need to tackle each of these other things by section (side bar elements, file browser subsections, etc).

**Comment by marthacryan** [comment]
Just rebased off of master with the new lumino updates to aria roles in tab bars (https://github.com/jupyterlab/lumino/pull/132). @manfromjupyter Maybe you could give it one last look to make sure everything is still working? As far as I understand, the only change should be some added aria info to tab bars / tabs (see referenced lumino PR for more info) and changing the role on tab lists from `navigation` to `tablist`. 

There's still one aria-label that isn't being translated, but we agreed that it shouldn't block on this PR at the moment.

**Comment by manfromjupyter** [comment]
@marthacryan sorry I'm so late to this. Saw it but just been overly busy. Ran it on binder and hitting an error. I'm not even using AWS so think it might be referring to binder itself, I guess?
getting binder error

**Comment by marthacryan** [comment]
@manfromjupyter Hmm I'm not seeing that errpr, maybe it was just a temporary issue with AWS?

**Comment by manfromjupyter** [comment]
@marthacryan I don't know what binder is doing differently in the last month because I've used my same VM since we started this project but I'm only get this issue now. Tried it on both a Linux and Windows VM within AWS and on multiple devices/IPs, it just does not like AWS for whatever reason. Googled for solutions and there's nothing; so must be a new issue. Sorry for this, will have to test it at the end of the day on a device on a different service.

**Comment by tomfried** [comment]
@marthacryan so I don't have a JAWS license for the computer I tested this on so I checked it all manually and it looks good behind 2-3 minor changes I include below, but if I missed a `<header>` or something burried deep, just know that's why hahaha.

I confirm all changes are in there that needed addressing, just 2-3 little tweaks here.
- Need `aria-orientation` set to `"vertical"` not `"horizontal"` for the sidebars. It needs to be `horizontal` for the `data-orientation` because that's referring to the way the 2 side bars, 2 side bar panel regions and notebook/main region is rendered in the middle (my guess); they are all horizontal to one another.
- Need the role="complementary" to the sidebar outer elements
- The right side bar should have the "navigation" scrapped from the end of it's `aria-label`, so just make it say `"alternate sidebar"` not `"alternate sidebar navigation"`. Screenreader will say "alternate sidebar tablist" because of that role. So don't need to manually add that piece of additional context to the user to differentiate it from the alternate sidebar region it lives under; the role will handle that).
left-side-bar-fixes
right-side-bar-fixes

**Tested and confirm everything else is golden**
- Editor nav is good
- Center/notebook pane set as main role is good
- File browser nav is good
- Actual Header (banner) section is good
- Sidebar tablists are good
- Removal of those `<header>` tags in the sidebar panels are gone (thank you for that). Manually checked but pretty sure there are none left
- Region roles set for the sidebar panels looks good

**Just a note:** Seems a little verbose with the large number of roles we have in here but think it's fine 😂 . I'm sure it will provide much needed context to a lot of these elements for screenreader users... which they probably care about more than saving clicks if jumping landmark to landmark.

**Comment by manfromjupyter** [comment]
@marthacryan so I don't have a JAWS license for the computer I tested this on so I checked it all manually and it looks good behind 2-3 minor changes I include below, but if I missed a `<header>` or something burried deep, just know that's why hahaha.

I confirm all changes are in there that needed addressing, just 2-3 little tweaks here.
- Need `aria-orientation` set to `"vertical"` not `"horizontal"` for the sidebars. It needs to be `horizontal` for the `data-orientation` because that's referring to the way the 2 side bars, 2 side bar panel regions and notebook/main region is rendered in the middle (my guess); they are all horizontal to one another.
- Need the role="complementary" to the sidebar outer elements
- The right side bar should have the "navigation" scrapped from the end of it's `aria-label`, so just make it say `"alternate sidebar"` not `"alternate sidebar navigation"`. Screenreader will say "alternate sidebar tablist" because of that role. So don't need to manually add that piece of additional context to the user to differentiate it from the alternate sidebar region it lives under; the role will handle that).
left-side-bar-fixes
right-side-bar-fixes

**Tested and confirm everything else is golden**
- Editor nav is good
- Center/notebook pane set as main role is good
- File browser nav is good
- Actual Header (banner) section is good
- Sidebar tablists are good
- Removal of those `<header>` tags in the sidebar panels are gone (thank you for that). Manually checked but pretty sure there are none left
- Region roles set for the sidebar panels looks good

**Just a note:** Seems a little verbose with the large number of roles we have in here but think it's fine 😂 . I'm sure it will provide much needed context to a lot of these elements for screenreader users... which they probably care about more than saving clicks if jumping landmark to landmark.

**Comment by marthacryan** [comment]
@manfromjupyter Just pushed those changes, thank you for checking it out!

**Review by manfromjupyter — APPROVED** [review]

**Review by blink1073 — APPROVED** [review]
Fantastic, thanks @marthacryan and @manfromjupyter! :heart:

**Pull request merged**